### PR TITLE
Use (implicit_transitive_deps false) in CI

### PR DESCRIPTION
**Pull request by dra27**
Split off from #4770 - won't build until that's merged!

### CODE REVIEW

**Comment by dra27** [comment]
We'll err on the side of caution and commit with `(implicit_transitive_deps true)` but in CI we'll alter it to be `(implicit_transitive_deps false)`. Our dependencies change infrequently enough that it shouldn't be an issue that CI, rather than development mode, is where this gets checked.

**Review by kit-ty-kate — COMMENTED** [review]
Looks decent as CI check. The only part I’m not too fond of is the `(re_export ...)`. If i understand what they do correctly, could we maybe add a TODO comment to those new additions to remind us to remove them when https://github.com/ocaml/ocaml/pull/11989 is merged and we use 5.1 in CI?

File: src/client/dune [hunk 1]
```diff
@@ -3,7 +3,7 @@
   (public_name opam-client)
   (synopsis    "OCaml Package Manager client and CLI library")
   (modules     (:standard \ opamMain get_git_version))
-  (libraries   opam-state opam-solver opam-repository re base64 cmdliner)
+  (libraries   opam-state opam-solver (re_export opam-repository) re base64 cmdliner)
```
Comment (kit-ty-kate) [inline src/client/dune:6]:
```suggestion
  ; TODO: Remove (re_export ...) when CI uses the OCaml version that includes https://github.com/ocaml/ocaml/pull/11989
  (libraries   opam-state opam-solver (re_export opam-repository) re base64 cmdliner)
```

File: src/core/dune [hunk 2]
```diff
@@ -2,7 +2,7 @@
   (name        opam_core)
   (public_name opam-core)
   (synopsis    "OCaml Package Manager core internal stdlib")
-  (libraries   re ocamlgraph unix bigarray sha jsonm
+  (libraries   re (re_export ocamlgraph) unix bigarray sha jsonm uutf
```
Comment (kit-ty-kate) [inline src/core/dune:5]:
```suggestion
  ; TODO: Remove (re_export ...) when CI uses the OCaml version that includes https://github.com/ocaml/ocaml/pull/11989
  (libraries   re (re_export ocamlgraph) unix bigarray sha jsonm uutf
```

File: src/format/dune [hunk 3]
```diff
@@ -2,7 +2,7 @@
   (name        opam_format)
   (public_name opam-format)
   (synopsis    "OCaml Package Manager file format handling library")
-  (libraries   opam-core opam-file-format re)
+  (libraries   (re_export opam-core) (re_export opam-file-format) re)
```
Comment (kit-ty-kate) [inline src/format/dune:5]:
```suggestion
  ; TODO: Remove (re_export ...) when CI uses the OCaml version that includes https://github.com/ocaml/ocaml/pull/11989
  (libraries   (re_export opam-core) (re_export opam-file-format) re)
```

File: src/repository/dune [hunk 4]
```diff
@@ -2,7 +2,7 @@
   (name        opam_repository)
   (public_name opam-repository)
   (synopsis    "OCaml Package Manager remote repository handling library")
-  (libraries   opam-format)
+  (libraries   (re_export opam-format))
```
Comment (kit-ty-kate) [inline src/repository/dune:5]:
```suggestion
  ; TODO: Remove (re_export ...) when CI uses the OCaml version that includes https://github.com/ocaml/ocaml/pull/11989
  (libraries   (re_export opam-format))
```

File: src/solver/dune [hunk 5]
```diff
@@ -2,7 +2,7 @@
   (name        opam_solver)
   (public_name opam-solver)
   (synopsis    "OCaml Package Manager solver interaction library")
-  (libraries   opam-format cudf dose3.algo re opam-0install-cudf
+  (libraries   (re_export opam-format) (re_export cudf) (re_export dose3.common) (re_export dose3.algo) re opam-0install-cudf
```
Comment (kit-ty-kate) [inline src/solver/dune:5]:
```suggestion
  ; TODO: Remove (re_export ...) when CI uses the OCaml version that includes https://github.com/ocaml/ocaml/pull/11989
  (libraries   (re_export opam-format) (re_export cudf) (re_export dose3.common) (re_export dose3.algo) re opam-0install-cudf
```

File: src/tools/dune [hunk 6]
```diff
@@ -3,7 +3,7 @@
   (public_name  opam-admin.top)
   (synopsis     "OCaml Package Manager admin toplevel")
   (modules      opam_admin_top)
-  (libraries    opam-client opam-file-format compiler-libs.toplevel re)
+  (libraries    opam-client opam-file-format (re_export compiler-libs.toplevel) re)
```
Comment (kit-ty-kate) [inline src/tools/dune:6]:
```suggestion
  ; TODO: Remove (re_export ...) when CI uses the OCaml version that includes https://github.com/ocaml/ocaml/pull/11989
  (libraries    opam-client opam-file-format (re_export compiler-libs.toplevel) re)
```

**Review by kit-ty-kate — COMMENTED** [review]
Something like that for example

**Review by kit-ty-kate — APPROVED** [review]

**Review by rjbou — APPROVED** [review]

**Comment by kit-ty-kate** [comment]
Thanks!

**Comment by dra27** [comment]
Thanks for adding the comments, @kit-ty-kate!

**Pull request merged**